File: dwcj-engine/src/test/java/org/dwcj/component/html/elements/AnchorTest.java
package org.dwcj.component.html.elements;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.fail;

import org.dwcj.component.element.PropertyDescriptorTester;
import org.junit.jupiter.api.Test;

class AnchorTest {

  @Test
  void shouldConstructWithHrefTextTarget() {
    String href = "https://example.com";
    String text = "Example Link";
    String target = "_blank";

    Anchor anchor = new Anchor(href, text, target);
    assertNotNull(anchor);
    assertEquals(href, anchor.getHref());
    assertEquals(text, anchor.getText());
    assertEquals(target, anchor.getTarget());
  }

  @Test
  void shouldConstructWithHrefText() {
    String href = "https://example.com";
    String text = "Example Link";

    Anchor anchor = new Anchor(href, text);
    assertNotNull(anchor);
    assertEquals(href, anchor.getHref());
    assertEquals(text, anchor.getText());
  }

  @Test
  void shouldConstructWithHrefAndComponents() {
    Div firstMock = new Div();
    Div secondMock = new Div();

    String href = "https://example.com";

    Anchor anchor = new Anchor(href, firstMock, secondMock);
    assertNotNull(anchor);
    assertEquals(href, anchor.getHref());
    assertEquals(2, anchor.getComponents().size());
  }

  @Test
  void shouldSetGetProperties() {
    Anchor component = new Anchor();

    try {
      PropertyDescriptorTester.run(Anchor.class, component);
    } catch (Exception e) {
      fail("PropertyDescriptor test failed: " + e.getMessage());
    }
  }

  @Test
  void setTargetShouldThrowExceptionWhenNull() {
    Anchor component = new Anchor();
    assertThrows(IllegalArgumentException.class, () -> {
      component.setTarget(null);
    });
  }

  @Test
  void setHrefShouldThrowExceptionWhenNull() {
    Anchor component = new Anchor();
    assertThrows(IllegalArgumentException.class, () -> {
      component.setUrl(null);
    });
  }
}

File: webforj-kotlin/src/test/kotlin/com/webforj/kotlin/dsl/component/html/elements/AnchorTest.kt
package com.webforj.kotlin.dsl.component.html.elements

import com.webforj.component.html.elements.Div
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

class AnchorTest {
    lateinit var root: Div

    @BeforeEach
    fun setUp() {
        root = Div()
    }

    @Test
    fun shouldCreateAnchorWithHref() {
        val expected = "href"
        val anchor = root.anchor(expected)
        assertEquals(expected, anchor.href)
        assertEquals("", anchor.text)
        assertEquals("_self", anchor.target)
        assertTrue { root.hasComponent(anchor) }
    }

    @Test
    fun shouldCreateAnchorWithText() {
        val expected = "text"
        val anchor = root.anchor(text = expected)
        assertEquals("", anchor.href)
        assertEquals(expected, anchor.text)
        assertEquals("_self", anchor.target)
        assertTrue { root.hasComponent(anchor) }
    }

    @Test
    fun shouldCreateAnchorWithTarget() {
        val expected = "target"
        val anchor = root.anchor(target = expected)
        assertEquals("", anchor.href)
        assertEquals("", anchor.text)
        assertEquals(expected, anchor.target)
        assertTrue { root.hasComponent(anchor) }
    }

    @Test
    fun shouldCreateAnchorWithHrefAndText() {
        val expectedHref = "href"
        val expectedText = "text"
        val anchor = root.anchor(expectedHref, expectedText)
        assertEquals(expectedHref, anchor.href)
        assertEquals(expectedText, anchor.text)
        assertEquals("_self", anchor.target)
        assertTrue { root.hasComponent(anchor) }
    }

    @Test
    fun shouldCreateAnchorWithHrefAndTarget() {
        val expectedHref = "href"
        val expectedTarget = "target"
        val anchor = root.anchor(expectedHref, target = expectedTarget)
        assertEquals(expectedHref, anchor.href)
        assertEquals("", anchor.text)
        assertEquals(expectedTarget, anchor.target)
        assertTrue { root.hasComponent(anchor) }
    }

    @Test
    fun shouldCreateAnchorWithTextAndTarget() {
        val expectedText = "text"
        val expectedTarget = "target"
        val anchor = root.anchor(text = expectedText, target = expectedTarget)
        assertEquals("", anchor.href)
        assertEquals(expectedText, anchor.text)
        assertEquals(expectedTarget, anchor.target)
        assertTrue { root.hasComponent(anchor) }
    }

    @Test
    fun shouldCreateAnchorWithHrefAndTextAndTarget() {
        val expectedHref = "href"
        val expectedText = "text"
        val expectedTarget = "target"
        val anchor = root.anchor(expectedHref, expectedText, expectedTarget)
        assertEquals(expectedHref, anchor.href)
        assertEquals(expectedText, anchor.text)
        assertEquals(expectedTarget, anchor.target)
        assertTrue { root.hasComponent(anchor) }
    }

}
